### PR TITLE
Add various workflows!

### PR DESCRIPTION
**Pull request by ajbt200128**
Added GHAs to: test/lint before PRs, and publish releases on tags.

Updated PR template

Tokens we need to set:

- `SEMGREP_APP_TOKEN` not sure where this is
- `OPEN_VSX_TOKEN` 1password -> Open VSX PAT
- `VS_MARKETPLACE_TOKEN` 1password -> semgrep-vscode Azure Devops publish token

### CODE REVIEW

File: .github/workflows/lint.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      # pre-commit is a Python script, this speedup things from Xmin to Ymin?
+      - uses: actions/setup-python@v4
+        with:
+          python-version: "3.11"
+          # This caches the pip installed dependencies that pre-commit uses.
+          # Otherwise it will need to reinstall everytime.
+          # TODO? who will need to reinstall everytime? github action? How much does this
+          # cache directive speedup this workflow?
+          cache: pip
+          cache-dependency-path: .github/workflows/lint.yaml
```
Comment (spencerdrak) [inline .github/workflows/lint.yaml:30]: I don't know if this should be a reference to another workflow, or at least, I don't understand what this is meant to do - https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md this uses the `cache-dependency-path` field differently.

File: .github/workflows/lint.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      # This is why sometimes pre-commit passes locally but fails in CI, for the same PR.
+      - uses: pre-commit/action@v3.0.0
+  github-actions:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: reviewdog/action-actionlint@v1
+        with:
+          fail_on_error: true
+          actionlint_flags: -ignore 'workflow command "set-output" was deprecated. ".+"'
```
Comment (spencerdrak) [inline .github/workflows/lint.yaml:43]: How come this ignore is required? I'm not seeing `set-output` used here, so we may want to not ignore this bit.

File: .github/workflows/publish.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  push:
+    tags:
+      - "*"
+
+name: Deploy Extension
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (spencerdrak) [inline .github/workflows/publish.yaml:11]: Recommend upgrade to v3

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    branches:
+      - develop
+    paths-ignore:
+      - "**.md"
+
+jobs:
+  vsce-prepublish:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (spencerdrak) [inline .github/workflows/tests.yml:18]: recommend using v3

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+name: tests
+
+on:
+  workflow_dispatch:
+  pull_request:
+    paths-ignore:
+      - "**.md"
+  push:
+    branches:
+      - develop
```
Comment (spencerdrak) [inline .github/workflows/tests.yml:10]: is the default branch meant to be `develop` or `main` or `master`? I've seen a few references to each of these. Probly would've been caught when the secrets got added and we tried to run but worth cleaning up.

File: .github/workflows/lint.yaml [hunk 6]
```diff
@@ -0,0 +1,43 @@
+name: lint
+
+on:
+  workflow_dispatch:
+  pull_request:
+  push:
+    branches:
+      - develop
```
Comment (spencerdrak) [inline .github/workflows/lint.yaml:8]: Same here, this `develop` may cause issues.

File: .github/workflows/semgrep.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      - master
+    paths:
+      - .github/workflows/semgrep.yml
+  schedule:
+    # random HH:MM to avoid a load spike on GitHub Actions at 00:00
+    - cron: "21 11 * * *"
+jobs:
+  semgrep:
+    name: semgrep/ci
+    runs-on: ubuntu-20.04
```
Comment (spencerdrak) [inline .github/workflows/semgrep.yaml:17]: I'd recommend standardizing on a version of `ubuntu` runner unless they specifically require it

**Review by spencerdrak — CHANGES_REQUESTED** [review]
Overall looks good - a few comments:
- consider if any further testing of these workflows is needed - it appears most are stateless so that shouldn't be an issue, but just something to keep in mind. I see you have a `prepublish` step, but might recommend merging this with the regular release step for clarity and then using the `if: ` blocks on the step (instead of the job) level for a bit more clarity. Do the reusable workflows we rely on support any sort of dry-run? Might also be useful.
- Left a few comments on things I don't fully understand, but overall structure seems fine

File: .github/workflows/publish.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+      - run: npm ci
+      - name: Publish to Open VSX Registry
+        uses: HaaLeo/publish-vscode-extension@v1
+        id: publishToOpenVSX
+        with:
+          pat: ${{ secrets.OPEN_VSX_TOKEN }}
+      - name: Publish to Visual Studio Marketplace
+        uses: HaaLeo/publish-vscode-extension@v1
+        with:
+          pat: ${{ secrets.VS_MARKETPLACE_TOKEN }}
```
Comment (aryx) [inline .github/workflows/publish.yaml:24]: same here

**Review by aryx — APPROVED** [review]

File: .github/workflows/lint.yaml (hunk 1 above)
Comment (ajbt200128) [reply .github/workflows/lint.yaml:30]: Hmm copied this from the semgrep repo, I'll remove it

File: .github/workflows/lint.yaml (hunk 2 above)
Comment (ajbt200128) [reply .github/workflows/lint.yaml:43]: Good point

File: .github/workflows/tests.yml (hunk 5 above)
Comment (ajbt200128) [reply .github/workflows/tests.yml:10]: Copied this from the semgrep repo and missed this, good catch. Main branch is master

File: .github/workflows/lint.yaml (hunk 1 above)
Comment (ajbt200128) [reply .github/workflows/lint.yaml:30]: Me neither tbh, especially considering the comment (I copied) is also unsure

**Comment by ajbt200128** [comment]
> * consider if any further testing of these workflows is needed - it appears most are stateless so that shouldn't be an issue, but just something to keep in mind. I see you have a `prepublish` step, but might recommend merging this with the regular release step for clarity and then using the `if: ` blocks on the step (instead of the job) level for a bit more clarity. ;

The prepublish step isn't actually a prerelease step, it's what the vscode extension cli calls the test it runs before publishing, so I just kept the name the same

**Comment by spencerdrak** [comment]
@ajbt200128 yep, sounds good. I was more referring to how we ensure that the workflows will be successful when we do end up pushing that tag to publish this. I don't think it's a blocker, but something to consider. I see Pad stamped, and I've got no further issues either.

**Review by spencerdrak — APPROVED** [review]

**Pull request merged**